org.eclipse.xtext/src/org/eclipse/xtext/generator/trace/node/GeneratorNodeProcessor.xtend

```diff
@@ -66,59 +66,60 @@ class GeneratorNodeProcessor {
 	def Result process(IGeneratorNode root) {
 		val ctx = new Context => [
 			lines = newArrayList(new StringBuilder)
-			currentIndents = new ArrayDeque<String>()
+			currentIndents = new ArrayDeque
 			pendingIndent = true
 		]
 		doProcess(root, ctx)
 		return new Result(ctx.lines.join, ctx.currentRegion)
 	}
 
 	/**
-	 * Indent nodes apply indentation between newline and content of its children.
+	 * An indent node prepends indentation to each line of its children.
 	 */
 	protected def dispatch void doProcess(IndentNode node, Context ctx) {
-		// do nothing if the indent node is empty
-		if (node.children.empty) {
-			return
-		}
-		try {
-			ctx.currentIndents.push(node.indentationString)
-			ctx.pendingIndent = true
-			doProcessChildren(node, ctx)
-		} finally {
-			ctx.currentIndents.pop
+		if (node.indentImmediately && !ctx.pendingIndent) {
+			ctx.currentLine.append(node.indentationString)
+		}
```

Comment:
I did that to avoid too many special cases. Probably it makes sense to check whether there is at least one leaf node. I'll change this.